src/access/universal.ts

```diff
@@ -32,3 +61,317 @@ export interface Access {
+ *   functions in this module, it is possible that it has been set in a way that
+ *   prevents this function from reliably setting access, in which case it will
+ *   resolve to `null`.
+ * - It will only set access explicitly for the given Agent. In other words,
+ *   additional restrictions could be present that further restrict or loosen
+ *   what access the given Agent has in particular circumstances.
+ * - The provided access will only apply to the given Resource. In other words,
+ *   if the Resource is a Container, the configured Access may not apply to
+ *   contained Resources.
+ * - If the current user does not have permission to view or change access for
+ *   the given Resource, this function will resolve to `null`.
+ *
+ * Additionally, two caveats apply to users with a Pod server that uses WAC:
+ * - If the Resource did not have an ACL yet, a new one will be initialised.
+ *   This means that changes to the ACL of a parent Container can no longer
+ *   affect access people have to this Resource, although existing access will
+ *   be preserved.
+ * - Setting different values for `controlRead` and `controlWrite` is not
+ *   supported, and **will throw an error**. If you expect (some of) your users
+ *   to have Pods implementing WAC, be sure to pass the same value for both.
+ * - Denying access is not technically possible in WAC. Rather, what this does
+ *   is removing "allow" access.
+ *
+ * @param resourceUrl URL of the Resource you want to change the Agent's access to.
+ * @param webId WebID of the Agent you want to set access for.
+ * @param access What access permissions you want to set for the given Agent to the given Resource. Possible properties are `read`, `append`, `write`, `controlRead` and `controlWrite`: set to `true` to allow, to `false` to deny, or `undefined` to leave unchanged. Take note that `controlRead` and `controlWrite` can not have distinct values for a Pod server implementing Web Access Control; trying this will throw an error.
+ */
+export async function setAgentAccess(
+  resourceUrl: UrlString,
+  webId: WebId,
+  access: Access,
+  options = internal_defaultFetchOptions
+): Promise<Access | null> {
+  const resourceInfo = await getResourceInfoWithAcr(resourceUrl, options);
+  if (hasAccessibleAcr(resourceInfo)) {
+    const updatedResource = setAgentAccessAcp(resourceInfo, webId, access);
+    if (updatedResource) {
+      try {
+        await saveAcrFor(updatedResource, options);
+        return getAgentAccessAcp(updatedResource, webId);
+      } catch (e) {
+        return null;
+      }
+    }
+    return null;
+  }
+  if (hasAccessibleAcl(resourceInfo)) {
+    if (access.controlRead != access.controlWrite) {
+      throw new Error(
+        `When setting access for a Resource in a Pod implementing Web Access Control (i.e. [${getSourceIri(
+          resourceInfo
+        )}]), ` + "`controlRead` and `controlWrite` should have the same value."
+      );
+    }
```

Comment:
This is also tested for in the WAC part of the API. I guess testing it only once is sufficient, but I'm not sure which one is more relevant. I guess testing earlier is best, and we can update the underlying type definitions to reflect the fact that this test happened already.

Reply:
Yeah, I was wondering the same. I think there are two possible approaches:

- We change the WAC interface to not take `controlRead` and `controlWrite`, but just a single `control` property. After all, it doesn't support the difference anyway.
- We keep the WAC interface aligned with the universal one, with the requirement that `controlRead` and `controlWrite` are equal. In that case, the check in the WAC module is technically redundant, because TypeScript will enforce it, but I think it might be good to keep it there just in case we ever expose it to non-TS users, or if the TS interface ever gets problematic and we remove that check.

If we were starting from scratch, I think I'd lean to the former, but I'm not sure if I'm leaning towards that strong enough to outweigh the fact that the latter is already implemented. Either way changing it would be a non-breaking change, luckily :)

Reply:
Yeah, if we had approached it in a different order, I'd have done option 1, but now that the checks are both there, it's more work to refactor things for little benefit I think. I implemented the WAC check in addition to the TS check in the spirit of defensive programming, so I guess it doesn't hurt if it stays there.